tchannel/tornado/connection.py

```diff
@@ -701,3 +634,176 @@ def _request_timed_out(self, req_id, req_ttl, future):
         # while.
         future.set_exception(errors.TimeoutError())
         self._request_tombstones.add(req_id, req_ttl)
+
+
+class Reader(object):
+
+    def __init__(self, io_stream, capacity=64):
+        self.queue = tornado.queues.Queue(capacity)
+        self.filling = False
+        self.io_stream = io_stream
+
+    def fill(self):
+        self.filling = True
+
+        io_loop = IOLoop.current()
+
+        def keep_reading(f):
+            if f.exception():
+                self.filling = False
+                return log(f.exception())
+            # connect these two in the case when put blocks
+            self.queue.put(f.result())
+            io_loop.spawn_callback(self.fill)
+
+        io_loop.add_future(read_message(self.io_stream), keep_reading)
+
+    def get(self):
+        """Receive the next message off the wire.
+
+        :returns:
+            A Future that produces a Context object containing the next
+            message off the wire.
+        """
+        if self.filling is False:
+            self.fill()
+
+        return self.queue.get()
+
+
+class Writer(object):
+
+    def __init__(self, io_stream, capacity=64):
+        self.queue = tornado.queues.Queue(capacity)
+        self.draining = False
+        self.io_stream = io_stream
+        # Tracks message IDs for this connection.
+        self._id_sequence = 0
+
+    def drain(self):
+        self.draining = True
+
+        io_loop = IOLoop.current()
+
+        def on_write(f, done):
+            if f.exception():
+                log.error(f.exception())
+                done.set_exc_info(f.exc_info())
+            else:
+                done.set_result(f.result())
+
+            io_loop.spawn_callback(next_write)
+
+        def on_message(f):
+            if f.exception():
+                io_loop.spawn_callback(next_write)
+                log.error(f.exception())
+                return
+            message, done = f.result()
+            io_loop.add_future(
+                self.io_stream.write(message),
+                lambda f: on_write(f, done),
+            )
+
+        def next_write():
+            if self.io_stream.closed():
+                return
+
+            io_loop.add_future(self.queue.get(), on_message)
+
+        io_loop.spawn_callback(next_write)
+
+    def put(self, message):
+        """Enqueues the given message for writing to the wire.
+
+        The message must be small enough to fit in a single frame.
+        """
+        if self.draining is False:
+            self.drain()
+
+        return self._enqueue(message)
```

Comment:
It is getting waited on. `connection.write` returns a future that resolves once the last fragment has been written, and that's waited on in `_stream`.